### PR TITLE
New Adapter: ADXCG

### PR DESCRIPTION
**Pull request by adxcgcom**
Hey!

This PR adds the ADXCG adapter

Contact email of the adapter's maintainer: info@adxcg.com

### CODE REVIEW

File: adapters/adxcg/adxcg.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
+package adxcg
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/mxmCherry/openrtb/v14/openrtb2"
+	//"github.com/mxmCherry/openrtb"
```
Comment (SyntaxNode) [inline adapters/adxcg/adxcg.go:9]: Please remove the commented out import package. Please also be aware there is a PR to upgrade `mxmCherry/openrtb` to version 15 which will eventually be a conflict to be resolved. Should be simple, it's the import path.

File: adapters/adxcg/adxcg.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+// Builder builds a new instance of the Adxcg adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	bidder := &AdxcgAdapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+// MediaOptiAdapter implements the Bidder interface.
+type AdxcgAdapter struct {
```
Comment (SyntaxNode) [inline adapters/adxcg/adxcg.go:25]: We prefer the name `adapter`. No comment is necessary.

File: adapters/adxcg/adxcg.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		Body:    openRTBRequestJSON,
+		Headers: headers,
+	}
+	requestsToBidder = append(requestsToBidder, requestToBidder)
+
+	return requestsToBidder, errs
+}
+
+const unexpectedStatusCodeFormat = "" +
+	"Unexpected status code: %d. Run with request.debug = 1 for more info"
```
Comment (SyntaxNode) [inline adapters/adxcg/adxcg.go:57]: I'm curious why you include the ` "" +`?

File: adapters/adxcg/adxcg.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+		return nil, []error{err}
+	}
+
+	bidsCapacity := len(openRTBBidderResponse.SeatBid[0].Bid)
+	bidderResponse = adapters.NewBidderResponseWithBidsCapacity(bidsCapacity)
+	var typedBid *adapters.TypedBid
+	for _, seatBid := range openRTBBidderResponse.SeatBid {
+		for _, bid := range seatBid.Bid {
+			bid := bid // pin! -> https://github.com/kyoh86/scopelint#whats-this
+			//TODO this must return data from extension
```
Comment (SyntaxNode) [inline adapters/adxcg/adxcg.go:96]: Please resolve this TODO.

File: adapters/adxcg/adxcg.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType, error) {
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Native != nil {
+				return openrtb_ext.BidTypeNative, nil
+			} else if imp.Banner != nil {
+				return openrtb_ext.BidTypeBanner, nil
+			} else if imp.Video != nil {
+				return openrtb_ext.BidTypeVideo, nil
+			}
```
Comment (SyntaxNode) [inline adapters/adxcg/adxcg.go:122]: This is an anti-pattern. Please use the response from your bidding server to determine the bid type.

File: static/bidder-info/adxcg.yaml [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+  app:
+    mediaTypes:
+      - banner
+      - video
+      - native
+  site:
+    mediaTypes:
+      - banner
+      - video
+      - native
```
Comment (SyntaxNode) [inline static/bidder-info/adxcg.yaml:13]: Does this bidder have a vendor id from the [IAB GDPR Vendor List](https://vendor-list.consensu.org/v2/vendor-list.json)?

Comment (adxcgcom) [reply static/bidder-info/adxcg.yaml:13]: no

File: adapters/adxcg/adxcg.go (hunk 4 above)
Comment (adxcgcom) [reply adapters/adxcg/adxcg.go]: removed

File: adapters/adxcg/adxcg.go (hunk 5 above)
Comment (adxcgcom) [reply adapters/adxcg/adxcg.go:117]: Would like to ... but backend bidding servers don't have this property ready for Live enviornment.

File: adapters/adxcg/adxcg.go (hunk 3 above)
Comment (adxcgcom) [reply adapters/adxcg/adxcg.go]: cleaned up

File: adapters/adxcg/adxcg.go (hunk 2 above)
Comment (adxcgcom) [reply adapters/adxcg/adxcg.go]: done

File: adapters/adxcg/adxcg.go (hunk 1 above)
Comment (adxcgcom) [reply adapters/adxcg/adxcg.go]: done

File: adapters/adxcg/adxcg.go (hunk 5 above)
Comment (SyntaxNode) [reply adapters/adxcg/adxcg.go:117]: Gotcha. For now, please be sure the order in this code matches the order evaluated by your servers. For example, according to this code, native will always be bid on first. Incorrect bid type detection will cause rendering issues on the client.

File: adapters/adxcg/adxcg.go (hunk 1 above)
Comment (SyntaxNode) [reply adapters/adxcg/adxcg.go]: That was a very quick response. Perhaps you'll make it in before the `mxmCherry/openrtb`.

**Comment by SyntaxNode** [comment]
Looking good. Please open a documentation PR: https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#user-documentation

**Comment by adxcgcom** [comment]
Thx for quick review.
Here is documentation PR: https://github.com/prebid/prebid.github.io/pull/2861

File: adapters/adxcg/adxcg.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+	return bidderResponse, nil
+
+}
+
+func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType, error) {
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Native != nil {
+				return openrtb_ext.BidTypeNative, nil
```
Comment (bsardo) [inline adapters/adxcg/adxcg.go:112]: Could you please add `simple-native.json` and `simple-video.json` tests to achieve better test coverage in this function?

File: adapters/adxcg/adxcg.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+	if err := json.Unmarshal(bidderRawResponse.Body, &openRTBBidderResponse); err != nil {
+		return nil, []error{err}
+	}
+
+	bidsCapacity := len(openRTBBidderResponse.SeatBid[0].Bid)
+	bidderResponse = adapters.NewBidderResponseWithBidsCapacity(bidsCapacity)
+	var typedBid *adapters.TypedBid
+	for _, seatBid := range openRTBBidderResponse.SeatBid {
+		for _, bid := range seatBid.Bid {
+			bid := bid // pin! -> https://github.com/kyoh86/scopelint#whats-this
```
Comment (bsardo) [inline adapters/adxcg/adxcg.go:92]: Nitpick: I know older adapters have this exact line but I suggest changing this to something like `bidCopy := bid` and updating the reference when creating a `TypedBid` down below to reduce confusion. The pinned comment is still helpful.

**Review by bsardo — COMMENTED** [review]
This is looking good, just a couple of comments.

File: adapters/adxcg/adxcg.go (hunk 7 above)
Comment (josepowera) [reply adapters/adxcg/adxcg.go:112]: done

File: adapters/adxcg/adxcg.go (hunk 8 above)
Comment (josepowera) [reply adapters/adxcg/adxcg.go]: thx 4 review.. done.

Comment (adxcgcom) [reply adapters/adxcg/adxcg.go]: Thx 4 review. Done both.

**Review by bsardo — APPROVED** [review]
LGTM

**Comment by adxcgcom** [comment]
Added change from openrtb v14-> v15 to get together with now approved https://github.com/prebid/prebid-server/pull/1802

File: adapters/adxcg/adxcgtest/exemplary/simple-native.json [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+            }
+          ],
+          "cur": "USD"
+        }
+      }
+    }
+  ],
+  "expectedBidResponses": [
+    {
+      "bids": [
```
Comment (SyntaxNode) [inline adapters/adxcg/adxcgtest/exemplary/simple-native.json:57]: While you're updating the library, could you please add the `"currency": "USD",` assertion in `expectedBidResponses` for simple-native and simple-video, as you have here in simple-banner.

**Review by bsardo — APPROVED** [review]

**Review by SyntaxNode — APPROVED** [review]

**Pull request merged**